tiledb/sm/rest/rest_profile.cc

```diff
@@ -0,0 +1,289 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file implements class RestProfile.
+ */
+
+#include <iostream>
+
+#include "rest_profile.h"
+#include "tiledb/common/random/random_label.h"
+#include "tiledb/sm/misc/constants.h"
+
+using namespace tiledb::common;
+using namespace tiledb::common::filesystem;
+
+namespace tiledb::sm {
+
+/* ****************************** */
+/*       PARAMETER DEFAULTS       */
+/* ****************************** */
+
+const std::string RestProfile::DEFAULT_NAME{"default"};
+const std::string RestProfile::DEFAULT_PASSWORD{""};
+const std::string RestProfile::DEFAULT_PAYER_NAMESPACE{""};
+const std::string RestProfile::DEFAULT_TOKEN{""};
+const std::string RestProfile::DEFAULT_SERVER_ADDRESS{"https://api.tiledb.com"};
+const std::string RestProfile::DEFAULT_USERNAME{""};
+
+const std::map<std::string, std::string> default_values = {
+    std::make_pair("rest.password", RestProfile::DEFAULT_PASSWORD),
+    std::make_pair(
+        "rest.payer_namespace", RestProfile::DEFAULT_PAYER_NAMESPACE),
+    std::make_pair("rest.token", RestProfile::DEFAULT_TOKEN),
+    std::make_pair("rest.server_address", RestProfile::DEFAULT_SERVER_ADDRESS),
+    std::make_pair("rest.username", RestProfile::DEFAULT_USERNAME)};
+
+/* ****************************** */
+/*   CONSTRUCTORS & DESTRUCTORS   */
+/* ****************************** */
+
+RestProfile::RestProfile(const std::string& name, const std::string& homedir)
+    : version_(constants::rest_profile_version)
+    , name_(name)
+    , filepath_(homedir + ".tiledb/profiles.json")
+    , old_filepath_(homedir + ".tiledb/cloud.json")
+    , param_values_(default_values) {
+  /**
+   * Ensure the user's $HOME is found.
+   * There's an edge case in which `sudo` does not always preserve the path to
+   * `$HOME`. In this case, the home_directory() API does not throw, but instead
+   * returns `std::nullopt`. As such, we can check for a value in the returned
+   * `std::optional` path of the home_directory and throw an error to the user
+   * accordingly, so they may decide the proper course of action: set the
+   * $HOME path, or perhaps stop using `sudo`.
+   */
+  if (homedir.empty()) {
```

Comment:
I think this check + comment belong to the other Constructor that actually create the home directory.
It's confusing that this constructor receives `homedir` as an input argument and we throw an error about `$HOME`.